temporalio/worker/_workflow.py

```diff
@@ -344,13 +345,124 @@ async def _handle_activation(
+        )
+
+        # Find the workflow to process safe eviction unless safe eviction
+        # disabled
+        workflow = None
+        if not self._disable_safe_eviction:
+            workflow = self._running_workflows.get(act.run_id)
+
+        # Safe eviction...
+        if workflow:
+            # We have to wait on the deadlocked task if it is set. This is
+            # because eviction may be the result of a deadlocked workflow but
+            # we cannot safely evict until that task is done with its thread. We
+            # don't care what errors may have occurred. We intentionally wait
+            # forever which means a deadlocked task cannot be evicted and give
+            # its slot back.
+            if workflow.deadlocked_activation_task:
+                logger.debug(
+                    "Waiting for deadlocked task to complete on run %s", act.run_id
+                )
+                try:
+                    await workflow.deadlocked_activation_task
+                except:
+                    pass
+
+            # Process the activation to evict. It is very important that
+            # eviction complete successfully because this is the only way we can
+            # confirm the event loop was torn down gracefully and therefore no
+            # GC'ing of the tasks occurs (which can cause them to wake up in
+            # different threads). We will wait deadlock timeout amount (2s if
+            # enabled) before making it clear to users that eviction is being
+            # swallowed. Any error or timeout of eviction causes us to retry
+            # forever because something in users code is preventing eviction.
+            seen_fail = False
+            handle_eviction_task: Optional[asyncio.Future] = None
+            while True:
+                try:
+                    # We only create the eviction task if we haven't already or
+                    # it is done. This is because if it already is running and
+                    # timed out, it's still running (and holding on to a
+                    # thread). But if did complete running but failed with
+                    # another error, we want to re-create the task.
+                    if not handle_eviction_task or handle_eviction_task.done():
+                        handle_eviction_task = (
+                            asyncio.get_running_loop().run_in_executor(
+                                self._workflow_task_executor,
+                                workflow.instance.activate,
+                                act,
+                            )
+                        )
+                    await asyncio.wait_for(
+                        handle_eviction_task, self._deadlock_timeout_seconds
+                    )
+                    # Break if it succeeds
+                    break
+                except BaseException as err:
+                    # Only want to log and mark as could not evict once
+                    if not seen_fail:
```

Comment:
IMO it could be worth logging this repeatedly, every 1 minute or something, just so it's extra obvious to people they've got something stuck.

Not blocking tho.

Reply:
We've found either people just haven't enabled logging or they found this log. We haven't seen situations where the log was not seen by users due to age. Would rather not add a timer mechanism into this unless I have to.